rmw_fastrtps_cpp/src/rmw_node.cpp

```diff
@@ -72,6 +81,14 @@ rmw_create_node(
 rmw_ret_t
 rmw_destroy_node(rmw_node_t * node)
 {
+  RMW_CHECK_ARGUMENT_FOR_NULL(node, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    node,
+    node->implementation_identifier,
+    eprosima_fastrtps_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  assert(node->context != nullptr);
```

Comment:
This assert seems a bit incongruous with the checking we do in e.g. `rwm_create_node`.  That is, this crashes the program, while most other places we do checks like this we just return an error from the function.  Also, this is somewhat of a philosophical debate, but my view is that a library should never assert; that's an application-level decision.  I'll suggest something more like:

```
if (node->context != nullptr) {
  RCUTILS_SET_ERROR_MSG("Attempted to destroy invalid context");
  return RMW_RET_INVALID_ARGUMENT;
}
```

Reply:
> This assert seems a bit incongruous with the checking we do in e.g. `rmw_create_node`.

How so? `rmw_create_node()` checks for a zero-initialized `context` because that's one of the possible `context` states (i.e. right before initialization and after finalization). There's no correct program that may call `rmw_destroy_node()` on a half-initialized or even uninitialized `rmw_node_t` structure.

> Also, this is somewhat of a philosophical debate, but my view is that a library should never assert; that's an application-level decision. 

I disagree. To return an error (or to throw an exception) implies there is a path to deal with the situation in a sane way. If you have hit UB territory, either your program is logically incorrect or it has managed to corrupt itself. There's no way to handle the situation other than to crash fast. That applies to applications and libraries the same.

Reply:
Which reminds me that we should document that after `rmw_destroy_node()`, using the pointer is UB.

Reply:
@hidmic and I discussed this at some length yesterday.  The pretty short summary of the discussion goes something like this:

* The caller can do arbitrary things to make this UB.  Like the following:

```
auto node = rmw_create_node(...);
memset(node, 0xFF, sizeof(rmw_node_t));
rmw_destroy_node(node);
```

There is nothing that we can do about cases like this; it's clearly an abuse of the API/library, and there is nothing we can reasonably check to defend against this.

* Accessing a nullptr here is really just another case of UB.  Take the above example and use a 0x0 instead:

```
auto node = rmw_create_node(...);
memset(node, 0x0, sizeof(rmw_node_t));
rmw_destroy_node(node);
```

And you have the same problem.  Thus, there's no real reason to treat `nullptr` specially here.

* The only other problem that we could potentially guard against is the user somehow intentionally setting the pointer to nullptr.  We could have a check for that, but its pretty unlikely that the user has done this on purpose.  Thus, for all practical purposes, this is UB too.

The conclusion from all of the above is that in all reasonable cases, it is UB, and thus we shouldn't have any checking here.  That's why @hidmic removed both the assert and the check from this code.

Reply:
> We could have a check for that, but its pretty unlikely that the user has done this on purpose.

For that matter, I'd say it's equally likely than `static_cast<rmw_node_impl_t *>(0xFF)`. In both cases, the user is departing from what's documented and corrupting the structure with pointers that cannot be de-referenced.